File: go/crud/wait.go
package crud

import (
	"time"

	"github.com/fraym/freym-api/go/proto/crud/deliverypb"
)

type Wait struct {
	ConditionFilter  *Filter
	ConditionTimeout time.Duration
}

func (w *Wait) toDeliveryWait() *deliverypb.DataWait {
	if w == nil || w.ConditionFilter == nil {
		return nil
	}

	return deliverypb.DataWait_builder{
		ConditionFilter: w.ConditionFilter.toProtobufFilter(),
		Timeout:         w.ConditionTimeout.Milliseconds(),
	}.Build()
}

type ListWait struct {
	Condition        *ListWaitCondition
	ConditionTimeout time.Duration
}

type ListWaitCondition struct {
	Operation string
	Value     int64
}

func (w *ListWait) toDeliveryListWait() *deliverypb.DataListWait {
	if w == nil || w.Condition == nil {
		return nil
	}

	return deliverypb.DataListWait_builder{
		Condition: deliverypb.DataListWaitCondition_builder{
			Operation: w.Condition.Operation,
			Value:     w.Condition.Value,
		}.Build(),
		Timeout: w.ConditionTimeout.Milliseconds(),
	}.Build()
}

File: js/crud/client/wait.ts
import { DataListWait, DataWait } from "@fraym/proto/dist/index.freym.crud.delivery";
import { Filter, getProtobufDataFilter } from "./filter";

export interface Wait {
    timeout?: number;
    conditionFilter: Filter;
}

export const getProtobufDataWait = (wait?: Wait): DataWait | undefined => {
    if (!wait) {
        return undefined;
    }

    return {
        conditionFilter: getProtobufDataFilter(wait.conditionFilter),
        timeout: (wait.timeout ?? 0).toString(),
    };
};

export interface ListWait {
    timeout?: number;
    condition: DataListWaitCondition;
}

export interface DataListWaitCondition {
    operation: string;
    value: number;
}

export const getProtobufDataListWait = (wait?: ListWait): DataListWait | undefined => {
    if (!wait) {
        return undefined;
    }

    return {
        condition: {
            operation: wait.condition.operation,
            value: wait.condition.value.toString(),
        },
        timeout: (wait.timeout ?? 0).toString(),
    };
};
